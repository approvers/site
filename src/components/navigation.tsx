import {
  Button,
  Container,
  Drawer,
  DrawerBody,
  DrawerCloseButton,
  DrawerContent,
  DrawerHeader,
  DrawerOverlay,
  Link,
  VStack,
  useDisclosure,
} from "@chakra-ui/react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import NextLink from "next/link";
import { faBars } from "@fortawesome/free-solid-svg-icons";
import { useRef } from "react";

export interface LinksProps {
  links: readonly Readonly<{
    name: string;
    url: string;
  }>[];
}

const Links = ({ links }: LinksProps): JSX.Element => (
  <VStack>
    {links.map(({ name, url }) => (
      <Container key={name}>
        <NextLink href={url} passHref>
          <Link>
            <Button minW="100%">{name}</Button>
          </Link>
        </NextLink>
      </Container>
    ))}
  </VStack>
);

export const Navigation = (props: LinksProps): JSX.Element => {
  const { isOpen, onOpen, onClose } = useDisclosure();
  const btnRef = useRef(null);

  return (
    <>
      <Button
        ref={btnRef}
        leftIcon={<FontAwesomeIcon icon={faBars} />}
        onClick={onOpen}
        pos="fixed"
        m={8}
        bottom={0}
        zIndex={100}
      >
        メニュー
      </Button>
      <Drawer isOpen={isOpen} placement="left" onClose={onClose} finalFocusRef={btnRef}>
        <DrawerOverlay />
        <DrawerContent>
          <DrawerCloseButton />
          <DrawerHeader borderBottomWidth="1px">Approvers</DrawerHeader>
          <DrawerBody>
            <Links {...props} />
          </DrawerBody>
        </DrawerContent>
      </Drawer>
    </>
  );
};
