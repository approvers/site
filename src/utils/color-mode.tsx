"use client";

import type { IconButtonProps, SpanProps } from "@chakra-ui/react";
import { ClientOnly, IconButton, Skeleton, Span } from "@chakra-ui/react";
import { ThemeProvider, useTheme } from "next-themes";
import type { ThemeProviderProps } from "next-themes";
import * as React from "react";
import { LuMoon, LuSun } from "react-icons/lu";

export interface ColorModeProviderProps extends ThemeProviderProps {}

export function ColorModeProvider(props: ColorModeProviderProps) {
  return <ThemeProvider attribute="class" disableTransitionOnChange {...props} />;
}

export type ColorMode = "light" | "dark";

export interface UseColorModeReturn {
  colorMode: ColorMode;
  setColorMode: (colorMode: ColorMode) => void;
  toggleColorMode: () => void;
}

export function useColorMode(): UseColorModeReturn {
  const { resolvedTheme, setTheme, forcedTheme } = useTheme();
  const colorMode = forcedTheme || resolvedTheme;
  const toggleColorMode = () => {
    setTheme(resolvedTheme === "dark" ? "light" : "dark");
  };
  return {
    colorMode: colorMode as ColorMode,
    setColorMode: setTheme,
    toggleColorMode,
  };
}

export function useColorModeValue<T>(light: T, dark: T) {
  const { colorMode } = useColorMode();
  return colorMode === "dark" ? dark : light;
}

export function ColorModeIcon() {
  const { colorMode } = useColorMode();
  return colorMode === "dark" ? <LuMoon /> : <LuSun />;
}

export function OppositeColorModeIcon() {
  const { colorMode } = useColorMode();
  return colorMode === "dark" ? <LuSun /> : <LuMoon />;
}

interface ColorModeButtonProps extends Omit<IconButtonProps, "aria-label"> {}

export const ColorModeButton = React.forwardRef<HTMLButtonElement, ColorModeButtonProps>(
  function ColorModeButton(props, ref) {
    const { toggleColorMode } = useColorMode();
    return (
      <ClientOnly fallback={<Skeleton boxSize="9" />}>
        <IconButton
          ref={ref}
          aria-label="Toggle color mode"
          onClick={toggleColorMode}
          size="sm"
          variant="ghost"
          {...props}
          css={{
            _icon: {
              width: "5",
              height: "5",
            },
          }}
        >
          <ColorModeIcon />
        </IconButton>
      </ClientOnly>
    );
  },
);

export const LightMode = React.forwardRef<HTMLSpanElement, SpanProps>(
  function LightMode(props, ref) {
    return (
      <Span
        className="chakra-theme light"
        ref={ref}
        display="contents"
        color="fg"
        colorPalette="gray"
        colorScheme="light"
        {...props}
      />
    );
  },
);

export const DarkMode = React.forwardRef<HTMLSpanElement, SpanProps>(function DarkMode(props, ref) {
  return (
    <Span
      className="chakra-theme dark"
      ref={ref}
      display="contents"
      color="fg"
      colorPalette="gray"
      colorScheme="dark"
      {...props}
    />
  );
});
