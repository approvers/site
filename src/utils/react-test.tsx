import reactTestRenderer, { ReactTestRenderer, TestRendererOptions } from "react-test-renderer";
import { ReactElement } from "react";

export const render = (
  ui: ReactElement,
  options?: TestRendererOptions | undefined,
): ReactTestRenderer => reactTestRenderer.create(ui, options);
