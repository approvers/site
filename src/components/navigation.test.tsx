import { expect, it } from "vitest";
import { Navigation } from "./navigation";
import { render } from "../utils/react-test";

it("renders correctly", () => {
  const tree = render(
    <Navigation
      links={[
        {
          name: "root",
          url: "/",
        },
        {
          name: "blog",
          url: "/blog",
        },

        {
          name: "contacts",
          url: "/contacts",
        },
        {
          name: "links",
          url: "/links",
        },
      ]}
    />,
  );
  expect(tree).toMatchSnapshot();
});
