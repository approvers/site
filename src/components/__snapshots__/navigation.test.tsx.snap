// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<button
  className="chakra-button css-zmq7l3"
  onClick={[Function]}
  type="button"
>
  <span
    className="chakra-button__icon css-1wh2kri"
  >
    <svg
      aria-hidden={true}
      className="svg-inline--fa fa-bars fa-w-14 "
      data-icon="bars"
      data-prefix="fas"
      focusable={false}
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </span>
  メニュー
</button>
`;
